dvc/cli.py

```diff
@@ -213,6 +313,8 @@ def get_main_parser():
     for cmd in COMMANDS:
         cmd.add_parser(subparsers, parent_parser)
 
+    parser.set_cmd_choices()
```

Comment:
Could we do this on-demand when we have to process the typo?

Reply:
The typos get processed in `parse_args()` so this could be set one time here:
```python
def parse_args(self, args=None, namespace=None):
    if not self._cmd_choices and not self._hidden_cmds:
        self.set_cmd_choices()
    
    if args is None:
        args = sys.argv[1:]
    else:
        args = list(args)
    ...
```

Reply:
@santos22 Do we need to cache it though? Unless I'm missing something, parse_args runs once, so we could just collect stuff we need once on error.

Reply:
No need to cache. I think instead of having `_cmd_choices` and `_hidden_cmds` be class properties they could be returned by the `set_cmd_choices()` method (we could rename it to `get_cmd_choices()`).

And maybe lines 158-185 could be put in a new method called `find_cmd_suggestions` to clean up the `parse_args` method.

So `parse_args` would look like:
```python
    def parse_args(self, args=None, namespace=None):
        cmd_choices, hidden_cmds = self.set_cmd_choices()

        if args is None:
            args = sys.argv[1:]
        else:
            args = list(args)

        self.find_cmd_suggestions(args, cmd_choices, hidden_cmds)

        # NOTE: overriding to provide a more granular help message.
        # E.g. `dvc plots diff --bad-flag` would result in a `dvc plots diff`
        # help message instead of generic `dvc` usage.
        args, argv = self.parse_known_args(args, namespace)
        if argv:
            msg = "unrecognized arguments: %s"
            self.error(msg % " ".join(argv), getattr(args, "func", None))
        return args
```